File: Leetcode/42-Trapping-Rain-Water/TrappingRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int total=0;
        if (height.size()<3) {
            return total;
        }
        int len = height.size();
        int left[len];
        int right[len];

        int max = height[0];
        for(int i=1; i<len; i++) {
            left[i] = std::max(max, height[i-1]);
            max = std::max(max, height[i]);
        }

        max = height[len-1];
        for (int i=len-2; i>=0; i--) {
            right[i] = std::max(max, height[i+1]);
            max = std::max(max, height[i]);
        }

        int sum;
        for (int i=1; i<len-1; i++) {
            sum = std::min(left[i], right[i]) - height[i];
            if (sum>0) {
                total += sum;
            }
        }
        return total;
    }
};

File: Leetcode/42-Trapping-Rain-Water/TrappingRainWater.go
func trap(height []int) int {
	total := 0
	if height == nil || len(height) < 3 {
		return total
	}

	length := len(height)
	left := make([]int, length)
	right := make([]int, length)

	m := height[0]
	for i := 1; i < length; i++ {
		left[i] = max(m, height[i-1])
		m = max(m, height[i])
	}

	m = height[length-1]
	for i := length - 2; i >= 0; i-- {
		right[i] = max(m, height[i+1])
		m = max(m, height[i])
	}

	sum := 0
	for i := 1; i < length-1; i++ {
		sum = min(left[i], right[i]) - height[i]
		if sum > 0 {
			total += sum
		}
	}
	return total
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

func min(i, j int) int {
	if i > j {
		return j
	}
	return i
}
